module/state_synchronization/getter.go

```diff
@@ -0,0 +1,340 @@
+	// in which case we should return the context error.
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
+	}
+
+	targetLogger.Debug().Msg("blob not found")
+
+	return nil, &BlobNotFoundError{target}
+}
+
+// getBlobs gets the given CIDs from the blobservice, reassembles the blobs, and deserializes the reassembled data into an object.
+//
+// blobs are fetched from the blobservice and sent over a blob channel as they arrive, to a separate goroutine, which performs the
+// deserialization in a streaming fashion.
+func (s *executionDataGetterImpl) getBlobs(ctx context.Context, cids []cid.Cid, logger zerolog.Logger) (interface{}, uint64, error) {
+	bcr, bs := blobs.OutgoingBlobChannel()
+
+	done := make(chan struct{})
+	var v interface{}
+	var deserializeErr error
+
+	// start deserialization goroutine
+	go func() {
+		defer close(done)
+		defer bcr.Close()
+
+		if v, deserializeErr = s.serializer.Deserialize(bcr); deserializeErr != nil {
+			logger.Debug().Err(deserializeErr).Msg("failed to deserialize execution data")
+		}
+	}()
+
+	retrieveErr := s.retrieveBlobs(ctx, bs, cids, logger)
+
+	<-done
+
+	if retrieveErr != nil && !errors.Is(retrieveErr, blobs.ErrClosedBlobChannel) {
+		return nil, 0, retrieveErr
+	}
+
+	if deserializeErr != nil {
+		return nil, 0, &MalformedDataError{deserializeErr}
+	}
+
+	// TODO: deserialization succeeds even if the blob channel reader has still has unconsumed data, meaning that a malicious actor
+	// could fill the blob tree with lots of unnecessary data by appending it at the end of the serialized data for each level.
+	// It's possible that we could detect this and fail deserialization using something like the following:
+	// https://github.com/onflow/flow-go/blob/bd5320719266b045ae2cac954f6a56e1e79560eb/engine/access/rest/handlers.go#L189-L193
+	// Eventually, we will need to implement validation logic on Verification nodes to slash this behavior.
+
+	return v, bcr.TotalBytesRead(), nil
+}
+
+func (s *executionDataGetterImpl) getExecutionDataRoot(
+	ctx context.Context,
+	blockID flow.Identifier,
+	rootID flow.Identifier,
+	blobGetter network.BlobGetter,
+	statusTracker StatusTracker,
+) (*ExecutionDataRoot, error) {
+	if err := statusTracker.StartTransfer(); err != nil {
+		return nil, fmt.Errorf("failed to track transfer start: %w", err)
+	}
+
+	blob, err := blobGetter.GetBlob(ctx, flow.IdToCid(rootID))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get execution data root blob: %w", err)
+	}
+
+	v, err := s.serializer.Deserialize(bytes.NewBuffer(blob.RawData()))
+	if err != nil {
+		return nil, fmt.Errorf("failed to deserialize execution data root blob: %w", err)
+	}
+
+	edRoot, ok := v.(*ExecutionDataRoot)
+	if !ok {
+		return nil, fmt.Errorf("execution data root blob is not an ExecutionDataRoot")
+	}
+
+	if edRoot.BlockID != blockID {
+		return nil, &MismatchedBlockIDError{blockID, edRoot.BlockID}
+	}
+
+	return edRoot, nil
+}
+
+func (s *executionDataGetterImpl) GetChunkExecutionDatas(
```

Comment:
Would this method being called multiple times?

Is this method a stateless method in a sense if called the second time, it won't "remember" the result it returned from the first time?

Reply:
It is not supposed to be called multiple times, but if it does get called again, it will load the data it already has on disk.

The way `BlobService` works is that it wraps the logic of checking the local blobstore before making a bitswap request into an interface, which you can call to transparently fetch needed blobs. So under the hood, when you call `BlobService.GetBlobs`, what actually happens is it first checks the blobstore if it already has any of those blobs, and if so, it sends those back immediately. It then sends bitswap requests for any ones it still needs.

So in the case `GetChunkExecutionDatas` gets called twice, the second time all of the data will be loaded from the local blobstore.

Reply:
Now a different question:

I had the impression that `Prune` is only for Getter, because BlobService would store too many data. But since BlobService is used by both Adder and Getter to add and get data by CID, the Adder would also have the same problem as Getter for storing too many data. So Adder will also need Prune as well, right?

Reply:
Potentially, it could use Prune, yes. But for the Adder (execution nodes), it can do something simpler which is just configure the underlying badger db with TTL expiry of all keys.

Pruning is probably the better solution, but using TTL is easier because we can avoid dealing with all of that stuff

Reply:
Using TTL might get the situation out of our control, especially in a case of running into a cadence bug. By the time, we fixed the bug, and past data might have all been pruned before being added?

TTL is usually used for cache. But our scenario here is not a cache. I think the fine-grained prune logic is necessary to guarantee the network has enough seeder for each block's execution data, right?

Reply:
Yeah I have same opinion as you theoretically. I just thought we could save some work and do this in a later PR. But now that I think about it more closely, I don't think we really save much work anyways.

To incorporate Pruning into the execution node, we need to prevent any new updates to the db while pruning happens, which means during the pruning it will not be able to store new Execution Data. It will be a bit of work to build in the synchronization mechanisms into the block executor logic to enable this.

Reply:
This also means that every once in a while, when pruning is happening, Execution Nodes will experience a pause in the block execution pipeline because Execution Data cannot be stored until the pruning is done. This would be infrequent (we can configure the pruning to only happen once per day or something, for example), but still, I wonder if this would be another reason to separate CID generation from the actual submission of the data blobs to badger / Bitswap as suggested [here](https://github.com/onflow/flow-go/pull/2292/files/7234bce41a91a71fad6244bcd3d70f5491cc905b#r849989908)?